File: ping/ping.go
package ping

// Copyright 2018 Comcast Cable Communications Management, LLC

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

// http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import "net"
import "golang.org/x/net/ipv4"
import "golang.org/x/net/ipv6"
import "golang.org/x/net/icmp"
import "time"
import "sync"
import "math"
import "connvitals/utils"

const IPV6_NETWORK_STRING = "ip6:ipv6-icmp";
const IPV4_NETWORK_STRING = "ip4:icmp"

/*
	A data structure that handles sending/receiving ICMP Echo ("ping") packets
*/
type Pinger struct {
	Host *net.IPAddr;
	IPv6 bool;
	Payload []byte;
	Connection *icmp.PacketConn;
	timestamps []time.Time;
	RTTS []time.Duration;
};

/*
	Builds a Pinger object, initializing the connection and setting the rtts to -1
*/
func New(host *net.IPAddr, IPv6 bool, payload int, numpings int) (pinger *Pinger, err error) {
	initialrtts := make([]time.Duration, numpings);
	for i := 0; i < numpings; i++ {
		initialrtts[i] = time.Duration(-1);
	}

	pinger = &Pinger{
	             host,
	             IPv6,
	             make([]byte, payload),
	             nil,
	             make([]time.Time, numpings),
	             initialrtts,
	         };

	if IPv6 {
		pinger.Connection, err = icmp.ListenPacket(IPV6_NETWORK_STRING, "::");
		if err != nil {
			return;
		}
		pinger.Connection.IPv6PacketConn().SetDeadline(time.Now().Add(2 * time.Second));
	} else {
		pinger.Connection, err = icmp.ListenPacket(IPV4_NETWORK_STRING, "0.0.0.0");
		if err != nil {
			return;
		}
		pinger.Connection.IPv4PacketConn().SetDeadline(time.Now().Add(2 * time.Second));
	}

	return;
}

/*
	Constructs an icmp packet to send along a connection
*/
func (pinger *Pinger) MkPacket(seqno int) (msg icmp.Message) {
	var typ icmp.Type;
	if pinger.IPv6 {
		typ = ipv6.ICMPTypeEchoRequest;
	} else {
		typ = ipv4.ICMPTypeEcho;
	}

	msg = icmp.Message{
		Type: typ,
		Code: 0,
		Body: &icmp.Echo{
			ID: 2,
			Seq: seqno,
			Data: pinger.Payload,
		},
	};

	return;
}

/*
	Sends a single packet on the Pinger's Connection, identified by the sequence number specified with seqno.
	It returns any errors generated from constructing the packet or sending it through the socket.
*/
func (pinger *Pinger) Send(seqno int) ( err error ) {
	pkt := pinger.MkPacket(seqno);
	var psh []byte = nil;
	if pinger.IPv6 {
		psh = icmp.IPv6PseudoHeader(pinger.Connection.LocalAddr().(*net.IPAddr).IP, pinger.Host.IP);
	}

	encodedPacket, err := pkt.Marshal(psh);
	if err != nil {
		return;
	}

	pinger.timestamps[seqno] = time.Now();
	if pinger.IPv6 {
		_, err = pinger.Connection.WriteTo(encodedPacket, pinger.Host);
	} else {
		_, err = pinger.Connection.WriteTo(encodedPacket, net.Addr(&net.IPAddr{IP: pinger.Host.IP.To4()}));
	}

	return;
}

/*
	Receives a single packet on the Pinger's Connection, and figures out what its sequence number is
	to calculate a round-trip time (rtt) for the packet. Returns errors caused by parsing messages.
*/
func (pinger *Pinger) Recv() (err error) {
	buf := make([]byte, 65536);
	var size int;
	var addr net.Addr;
	var msg *icmp.Message;

	// Wait for a ping from the host that we actually pinged
	for true {
		size, addr, err = pinger.Connection.ReadFrom(buf);
		if err != nil {
			err = nil; // This is almost certainly a timeout, so just ignore it (for now)
			return;
		} else if addr.(*net.IPAddr).IP.Equal(pinger.Host.IP) {
			var proto int;
			if pinger.IPv6 {
				proto = ipv6.ICMPTypeEchoRequest.Protocol();
			} else {
				proto = ipv4.ICMPTypeEcho.Protocol();
			}

			msg, err = icmp.ParseMessage(proto, buf[:size]);
			if err != nil {
				return;
			} else if (msg.Type == ipv4.ICMPTypeEchoReply || msg.Type == ipv6.ICMPTypeEchoReply ) && msg.Body.(*icmp.Echo).ID == 2 {
				break;
			}

		}
	}



	seqno := msg.Body.(*icmp.Echo).Seq;
	pinger.RTTS[seqno] = time.Since(pinger.timestamps[seqno]);
	return;
}



/*
	Pings a single host passed as an argument, and returns a result string that's ready for printing.
*/
func PingHost(host *net.IPAddr, IPv6 bool, numpings int, payload int) (min, avg, max, std, loss float64, err error) {
	pinger, err := New(host, IPv6, payload, numpings);
	if err != nil {
		return;
	}
	defer pinger.Connection.Close();

	var pool sync.WaitGroup;

	for i := 0; i < numpings; i++ {
		pool.Add(2);

		go func (seqno int) {
			defer pool.Done();
			err = pinger.Send(seqno);
			if err != nil {
				utils.Warn(err.Error());
			}
		}(i);

		go func () {
			defer pool.Done();
			err = pinger.Recv();
			if err != nil {
				utils.Warn(err.Error());
			}
		}();
	}

	// Wait for results
	pool.Wait();

	min = math.Inf(0);
	number_of_lost_packets := 0;
	for i := 0; i < numpings; i++ {
		rtt := float64(pinger.RTTS[i]) / float64(time.Millisecond);
		if rtt < 0 {
			number_of_lost_packets++;
			continue;
		}

		if min > rtt {
			min = rtt;
		}
		if max < rtt {
			max = rtt;
		}

		avg += rtt
	}

	avg /= float64(numpings - number_of_lost_packets);

	//Need to loop again once the average is found to get std
	for i := 0; i < numpings; i++ {
		rtt := float64(pinger.RTTS[i]) / float64(time.Millisecond);
		if rtt > 0 {
			std += math.Pow(rtt - avg, 2);
		}
	}
	std /= float64(numpings - 1 - number_of_lost_packets);
	std = math.Sqrt(std);

	// if all packets are lost, the values of min/avg/max/std are meaningless, so print this instead to avoid confusion
	if number_of_lost_packets >= numpings {
		return -1, -1, -1, -1, 100, nil;
	} else {
		loss = float64(number_of_lost_packets)/ float64(numpings) * 100.0;
	}

	return;
}

File: connvitals/ping.py
# Copyright 2018 Comcast Cable Communications Management, LLC

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

# http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
This module defines a class and utilities to manage icmp/icmpv6 echo requests
and replies to remote hosts.
"""

import socket
import struct
import time
import sys
import math
from . import utils
from . import icmp

def icmpParse(pkt: bytes, ipv6: bool) -> int:
	"""
	Parses an icmp packet, returning its sequence number.

	If the packet is found to be not an echo reply, this will
		immediately return -1, indicating that this packet
		should be disregarded.
	"""
	try:
		if ipv6:
			if pkt[0] == 129:
				return struct.unpack("!H", pkt[6:8])[0]
			return -1
		if pkt[20] == 0:
			return struct.unpack("!H", pkt[26:28])[0]
		return -1
	except (IndexError, struct.error):
		return -1

class Pinger():
	"""
	A data structure that handles icmp pings to a remote machine.
	"""
	def __init__(self, host: utils.Host, payload: bytes):
		"""
		Inializes a socket connection to the host on port 22, and returns a Pinger object
		referencing it.
		"""

		self.sock, self.icmpParse, self.mkPkt = None, None, None

		if host.family == socket.AF_INET6:
			self.sock = socket.socket(host.family, socket.SOCK_RAW, proto=socket.IPPROTO_ICMPV6)
			self.icmpParse = self._icmpv6Parse
			self.mkPkt = self._mkPkt6
		else:
			self.sock = socket.socket(host.family, socket.SOCK_RAW, proto=socket.IPPROTO_ICMP)
			self.icmpParse = self._icmpv4Parse
			self.mkPkt = self._mkPkt4

		self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

		self.sock.settimeout(1)
		self.payload = payload

		#Build a socket object
		self.host = host

		self.timestamps = {}

	def sendAll(self, num:int) -> utils.PingResult:
		"""
		Sends all pings sequentially in one thread
		"""
		for i in range(num):
			pkt = icmp.ICMPPkt(self.host, payload=struct.pack("!HH", 2, i))
			self.timestamps[i] = time.time()

			try:
				self.sock.sendto(bytes(pkt), (self.host.addr, 0))
			except Exception as e:
				utils.err("Network is unreachable... (%s)" % e)

		return self.recvAll(num)

	def recvAll(self, num:int) -> utils.PingResult:
		"""
		Recieves and parses all packets
		"""
		maxPacketLen, found = 100 + len(self.payload), 0
		pkts = [-1,]*num

		while found < num:

			try:
				pkt, addr = self.sock.recvfrom(maxPacketLen)
			except (socket.timeout, TimeoutError):
				# If we hit a socket timeout, we'll consider the packet dropped
				found += 1
				continue

			if addr[0] == self.host[0]:
				seqno = self.icmpParse(pkt)
				if seqno in range(len(pkts)):
					pkts[seqno] = time.time() - self.timestamps[seqno]
					found += 1

		# All packets collected; parse and return the results
		rtt, lost = [], 0
		for pkt in pkts:
			if pkt > 0:
				rtt.append(pkt*1000)
			else:
				lost += 1

		if lost == num:
			return utils.PingResult(-1, -1, -1, -1, 100.)

		try:
			avg = sum(rtt) / len(rtt)
			std = 0.
			for item in rtt:
				std += (avg - item) ** 2
			std /= len(rtt) - 1
			std = math.sqrt(std)
		except ZeroDivisionError:
			std = 0.

		return utils.PingResult(min(rtt), avg, max(rtt), std, lost/num * 100.0)

	def ping(self, seqno: int) -> float:
		"""
		Sends a single icmp packet to the remote host.
		Returns the round-trip time (in ms) between packet send and receipt
		or 0 if packet was not received.
		"""
		pkt = icmp.ICMPPkt(self.host, payload=struct.pack("!HH", 2, seqno))

		# I set time here so that rtt includes the device latency
		self.timestamps[seqno] = time.time()

		try:
			# ICMP has no notion of port numbers
			self.sock.sendto(bytes(pkt), (self.host.addr, 0))
		except Exception as e:
			#Sometimes, when the network is unreachable this will erroneously report that there's an
			#'invalid argument', which is impossible since the hostnames are coming straight from
			#`socket` itself
			raise Exception("Network is unreachable... (%s)" % e)
		return self.recv()

	@staticmethod
	def _icmpv4Parse(pkt: bytes) -> int:
		"""
		Attemtps to parse an icmpv4 packet, returning the sequence number if parsing succeds,
		or -1 otherwise.
		"""
		try:
			if pkt[20] == 0:
				return struct.unpack("!H", pkt[26:28])[0]
		except (IndexError, struct.error):
			pass
		return -1

	@staticmethod
	def _icmpv6Parse(pkt: bytes) -> int:
		"""
		Attemtps to parse an icmpv6 packet, returning the sequence number if parsing succeds,
		or -1 otherwise.
		"""
		try:
			if pkt[0] == 0x81:
				return struct.unpack("!H", pkt[6:8])[0]
		except (IndexError, struct.error):
			pass
		return -1

	def _mkPkt4(self, seqno: int) -> bytes:
		"""
		Contsructs and returns an ICMPv4 packet
		"""
		header = struct.pack("!BBHHH", 8, 0, 0, 2, seqno)
		checksum = self._checksum4(header + self.payload)
		return struct.pack("!BBHHH", 8, 0, checksum, 2, seqno) + self.payload

	def _mkPkt6(self, seqno: int) -> bytes:
		"""
		Contsructs and returns an ICMPv6 packet
		"""
		header = struct.pack("!BBHHH", 0x80, 0, 0, 2, seqno)
		checksum = self._checksum6(header)
		return struct.pack("!BBHHH", 0x80, 0, checksum, 2, seqno) + self.payload

	@staticmethod
	def _checksum4(pkt: bytes) -> int:
		"""
		calculates and returns the icmpv4 checksum of 'pkt'
		"""

		countTo = len(pkt) // 2 * 2
		total, count = 0, 0

		# Handle bytes in pairs (decoding as short ints)
		loByte, hiByte = 0, 0
		while count < countTo:
			if sys.byteorder == "little":
				loByte = pkt[count]
				hiByte = pkt[count + 1]
			else:
				loByte = pkt[count + 1]
				hiByte = pkt[count]
			total += hiByte * 256 + loByte
			count += 2

		# Handle last byte if applicable (odd-number of bytes)
		# Endianness should be irrelevant in this case
		if countTo < len(pkt): # Check for odd length
			total += pkt[len(pkt) - 1]

		total &= 0xffffffff # Truncate sum to 32 bits (a variance from ping.c, which
		                    # uses signed ints, but overflow is unlikely in ping)

		total = (total >> 16) + (total & 0xffff)    # Add high 16 bits to low 16 bits
		total += (total >> 16)                      # Add carry from above (if any)

		return socket.htons((~total) & 0xffff)

	def _checksum6(self, pkt: bytes) -> int:
		"""
		calculates and returns the icmpv6 checksum of pkt
		"""
		laddr = socket.inet_pton(self.host[1], self.sock.getsockname()[0])
		raddr = socket.inet_pton(*reversed(self.host))
		# IPv6 Pseudo-Header used for checksum calculation as specified by
		# RFC 2460 (https://tools.ieft.org/html/rfc2460)
		psh = laddr + raddr + struct.pack("!I", len(pkt)) + b'\x00\x00\x00:'
		# This last bit is the 4-byte-packed icmp6 protocol number (58 or 0xa3)


		total, packet = 0, psh+pkt

		# Sum all 2-byte words
		num_words = len(packet) // 2
		for chunk in struct.unpack("!%sH" % num_words, packet[0:num_words*2]):
			total += chunk

		# Add any left-over byte (for odd-length packets)
		if len(packet) % 2:
			total += ord(packet[-1]) << 8

		# Fold 32-bits into 16-bits
		total = (total >> 16) + (total & 0xffff)
		total += total >> 16
		return ~total + 0x10000 & 0xffff

	def recv(self) -> float:
		"""
		Recieves each ping sent.
		"""
		# If a packet is not an echo reply, icmpParse will give its seqno as -1
		# This lets us disregard packets from traceroutes immediately
		maxlen = 100+len(self.payload) #this should be enough to cover headers and whatnot

		while True:

			try:
				pkt, addr = self.sock.recvfrom(maxlen)
			except (socket.timeout, TimeoutError):
				return -1

			# The packet must have actually come from the host we pinged
			if addr[0] == self.host[0]:
				seqno = self.icmpParse(pkt)
				if seqno >= 0:
					return time.time() - self.timestamps[seqno]


	def __enter__(self) -> "Pinger":
		"""
		Context-managed instantiation
		"""
		return self

	def __exit__(self, exc_type, exc_value, traceback):
		"""
		Context-managed cleanup
		"""
		self.sock.close()

		if exc_type and exc_value:
			utils.error(exc_type("Unknown error occurred while pinging"))
			utils.error(exc_type(exc_value))
			if traceback:
				utils.warn(traceback)
